src/main/scala/beam/agentsim/infrastructure/ChargingNetworkManager.scala

```diff
@@ -0,0 +1,249 @@
+            }
+            .orElse {
+              vehiclesToCharge.update(vehicleId, ChargingVehicle(vehicle, agent, newTotalSession, currentSession))
+              None
+            }
+
+        case (id, energy) if energy <= 0 =>
+          log.warning(
+            "Vehicle {}  (primaryFuelLevel = {}) requires energy {} - which is less or equals zero",
+            vehicles(id),
+            vehicles(id).primaryFuelLevelInJoules,
+            energy
+          )
+          None
+      }.toVector
+
+      sender ! CompletionNotice(
+        triggerId,
+        if (tick < endOfSimulationTime)
+          scheduleTriggers :+ ScheduleTrigger(PlanningTimeOutTrigger(nextTick), self)
+        else {
+          // if we still have a BEV/PHEV that is connected to a charging point,
+          // we assume that they will charge until the end of the simulation and throwing events accordingly
+          val completeTriggers = scheduleTriggers ++ vehiclesToCharge.map {
+            case (_, cv) =>
+              ScheduleTrigger(
+                EndRefuelSessionTrigger(
+                  tick,
+                  cv.vehicle.getChargerConnectedTick(),
+                  cv.totalChargingSession.energy,
+                  cv.vehicle
+                ),
+                cv.agent
+              )
+          }
+          vehiclesToCharge.clear()
+          completeTriggers
+        }
+      )
+  }
+
+  private def replanHorizon(tick: Int): (Int, Map[Id[BeamVehicle], Double]) = {
+    val requiredPower = sitePowerManager.getPowerOverPlanningHorizon(vehicles)
+
+    val (bounds, nextTick) = if (gridConnectionEnabled) {
+      powerController.publishPowerOverPlanningHorizon(requiredPower, tick)
+      powerController.obtainPowerPhysicalBounds(tick)
+    } else {
+      powerController.defaultPowerPhysicalBounds(tick)
+    }
+    val requiredEnergyPerVehicle = sitePowerManager.replanHorizonAndGetChargingPlanPerVehicle(bounds, vehicles)
+
+    if (requiredEnergyPerVehicle.nonEmpty)
+      log.info("Required energy per vehicle: {}", requiredEnergyPerVehicle.mkString(","))
+
+    (nextTick, requiredEnergyPerVehicle)
+  }
+
+  private def endRefuelSessionTriggerMaybe(
+    vehicle: BeamVehicle,
+    tick: Int,
+    maxChargeDuration: Long,
+    currentSession: ChargingSession,
+    totalSession: ChargingSession
+  ): Option[EndRefuelSessionTrigger] = {
+    if (vehicle.primaryFuelLevelInJoules >= vehicle.beamVehicleType.primaryFuelCapacityInJoule) {
+      log.debug(
+        "Vehicle {} is fully charged. Scheduling EndRefuelSessionTrigger at {} with {} J delivered",
+        vehicle.id,
+        tick + currentSession.duration.toInt,
+        totalSession.energy
+      )
+      Some(
+        EndRefuelSessionTrigger(
+          tick + currentSession.duration.toInt,
+          vehicle.getChargerConnectedTick(),
+          totalSession.energy,
+          vehicle
+        )
+      )
+    } else if (currentSession.duration < maxChargeDuration) {
```

Comment:
Shouldn't this be the other way around? If the duration is more than the max duration then it should be ended, otherwise it is under max duration and can continue?

Reply:
it could not be more, because we limit it by `maxChargeDuration` here:
```
val (chargingDuration, providedEnergy) = vehicle.refuelingSessionDurationAndEnergyInJoules(Some(maxChargeDuration))
```

Reply:
I think part of my confusion was the usage of the word `over` - I had been in the mindset that it meant is it overcharged. Maybe rename the method to `isChargingCompleted`? Also, I saw your change from >= to == in the initial if. Maybe it's the worrier in me, but I would put a log or exception if it is > since it should never be...maybe a test to catch if we break that guarantee (if one isn't already there) - or both?